experimental/devices/hd44780/hd44780.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package hd44780 controls the Hitachi LCD display chipset HD-44780
+//
+// Datasheet
+//
+// https://www.sparkfun.com/datasheets/LCD/HD44780.pdf
+package hd44780
+
+import (
+	"fmt"
+	"time"
+
+	"periph.io/x/periph/conn"
+	"periph.io/x/periph/conn/gpio"
+)
+
+// lineTwo offset for the second line in the LCD buffer.
+const lineTwo = 0x40
+
+// Dev is the 4-bit addressing device for HD-44780
+type Dev struct {
+	// data pins
+	dataPins []gpio.PinOut
+
+	// register select pin
+	rsPin gpio.PinOut
+
+	// enable pin
+	enablePin gpio.PinOut
+}
+
+// New creates and initializes the LCD device
+//	data - references to data pins
+//	rs - rs pin
+//	e - strobe pin
+func New(data []gpio.PinOut, rs, e gpio.PinOut) (*Dev, error) {
+	if len(data) != 4 {
+		return nil, fmt.Errorf("expected 4 data pins, passed %s", len(data))
+	}
+	dev := &Dev{
+		dataPins:  data,
+		enablePin: e,
+		rsPin:     rs,
+	}
+	if err := dev.Init(); err != nil {
+		return nil, err
+	}
+	return dev, nil
+}
+
+// Init initializes the HC-44780 chipset, clears the screen buffer and moves cursor to the
+// home of screen (line 0, column 0).
+func (r *Dev) Init() error {
+	if err := r.clearBits(); err != nil {
+		return err
+	}
+
+	delayMs(15)
+
+	if err := r.rsPin.Out(gpio.Low); err != nil {
+		return err
+	}
+	if err := r.enablePin.Out(gpio.Low); err != nil {
+		return err
+	}
+
+	if err := r.bulkSendData(resetSequence, r.write4Bits); err != nil {
+		return err
+	}
+
+	if err := r.bulkSendData(initSequence, r.writeInstruction); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Halt will re-initialize the display.
+func (r *Dev) Halt() error {
+	return r.Init()
+}
+
+func (r *Dev) String() string {
+	return "HD44870, 4 bit mode"
+}
+
+// Cls clears the LCD screen
+func (r *Dev) Cls() error {
+	if err := r.writeInstruction(0x01); err != nil {
+		return err
+	}
+	delayMs(2)
+	return nil
+}
+
+// SetCursor positions the cursor
+//	line - screen line, 0-based
+//	column - column, 0-based
+func (r *Dev) SetCursor(line uint8, column uint8) error {
+	if err := r.writeInstruction(0x80 | (line*lineTwo + column)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Print the data string
+//	data string to display
+func (r *Dev) Print(data string) error {
+	for _, v := range []byte(data) {
+		if err := r.WriteChar(v); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PrintAt the data string at specified position
+//	data string to display
+func (r *Dev) PrintAt(row, column uint8, data string) error {
+	if err := r.SetCursor(row, column); err != nil {
+		return err
+	}
+	for _, v := range []byte(data) {
```

Comment:
That's incorrect. In Go, string represents an UTF-8 encoded string. You need to report unsupported characters and manually translate the supported ones into the code page.

This is particularly important as there's multiple chips with different code pages (HCD44780UA00 vs HCD44780UA02), and it's valuable to support custom characters.

I can't find in the datasheet a way to identify the code page of the chip. (?)

Reply:
AFAIK there's no way to tell which codepage the chip has. You can order the chip with your own codepage. 

Still you can upload up to 8 of custom character patterns and use those.

From this point, I don't see any value in working with runes here, `byte` seems to be a reasonable primitive - the end users may experiment with the encodings on the client side.

Reply:
Clarifying my idea:

```
type CodePage map[rune]string

var A00 = CodePage{...}
var A02 = CodePage{...}

func New(data []gpio.PinOut, rs, e gpio.PinOut, cp CodePage) (*Dev, error) {
...
```

Then here you can do:

```
for _, r := range data {
  b, ok := d.cp[r]
  if !ok {
     c = ' '
   }
   ....
```

Then in the doc, add:

// Unknown characters are printed as whitespace

Reply:
According to https://en.wikipedia.org/wiki/Hitachi_HD44780_LCD_controller
* HD44780UA00 - the standard (Japanese) version, which includes katakana characters and some Greek letters and mathematical symbols
* HD44780UA02 - a European version, which includes the complete set of Greek, Cyrillic and Western European characters (with diacritics)

Now on the datasheet: https://www.sparkfun.com/datasheets/LCD/HD44780.pdf, page 17 table 4 there is the list of stuff that is supported by this chipset.

Do seriously expect to have all the symbols enumerated in the code and that check to be implemented? I'm just curious.

Reply:
I do, but that's not a blocker for this CL.